.github/workflows/ruby.yml

```diff
@@ -0,0 +1,110 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+name: Ruby
+
+on:
+  push:
+    # paths:
+    #   - 'cpp/**'
+    #   - 'c_glib/**'
+    #   - 'ruby/**'
+  pull_request:
+    # paths:
+    #   - 'cpp/**'
+    #   - 'c_glib/**'
+    #   - 'ruby/**'
+
+jobs:
+
+  docker:
+    name: ${{ matrix.name }}
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        image:
+          - amd64-ubuntu-18.04-ruby
+        include:
+          - name: AMD64 Ubuntu 18.04 Ruby
+            image: amd64-ubuntu-18.04-ruby
+            ubuntu: 18.04
+            platform: ubuntu
+    env:
+      UBUNTU: ${{ matrix.ubuntu || 18.04 }}
+    steps:
+      - name: Checkout Arrow
+        uses: actions/checkout@v1
+        with:
+          submodules: true
+      - name: Docker Pull
+        shell: bash
+        run: |
+          docker-compose pull --ignore-pull-failures \
+            ${{ matrix.platform }}-cpp
+          docker-compose pull --ignore-pull-failures \
+            ${{ matrix.platform }}-c-glib
+          docker-compose pull --ignore-pull-failures \
+            ${{ matrix.platform }}-ruby
+      - name: Docker Build
+        shell: bash
+        run: |
+          docker-compose build ${{ matrix.platform }}-cpp
+          docker-compose build ${{ matrix.platform }}-c-glib
+          docker-compose build ${{ matrix.platform }}-ruby
+      - name: Docker Run
+        shell: bash
+        run: docker-compose run ${{ matrix.platform }}-ruby
+      - name: Docker Push
+        if: github.event_name == 'push' # && github.ref == 'master'
+        shell: bash
+        run: |
+          docker login \
+            -u ${{ secrets.DOCKERHUB_USER }} \
+            -p ${{ secrets.DOCKERHUB_TOKEN }}
+          docker-compose push ${{ matrix.platform }}-ruby
+
+  macos:
+    name: AMD64 MacOS 10.14 Ruby
+    runs-on: macos-10.14
+    strategy:
+      fail-fast: false
+    env:
+      ARROW_HOME: /usr/local
+      ARROW_ORC: OFF
+      ARROW_JEMALLOC: OFF
+      ARROW_BUILD_TESTS: ON
+    steps:
+      - name: Checkout Arrow
+        uses: actions/checkout@v1
+        with:
+          submodules: true
+      - name: Install Dependencies
+        shell: bash
+        run: |
+          brew bundle --file=cpp/Brewfile
+          brew bundle --file=c_glib/Brewfile
+          bundle install --gemfile ruby/red-arrow/Gemfile
+      - name: Build
+        shell: bash
+        run: |
+          ci/scripts/cpp_build.sh $(pwd) $(pwd)/cpp/build
+          ci/scripts/c_glib_build.sh $(pwd) $(pwd)/c_glib/build
```

Comment:
`C++` -> `C Glib` -> `Ruby` is the branch which we can indeed collapse to `C++` -> `C GLib & Ruby` because there isn't any other implementation depending on C Glib, just Ruby. Once we have more C Glib dependent impls, then we can expand the branch again to not trigger e.g. Lua builders for pure Ruby changes.